lib/Sema/CSGen.cpp

```diff
@@ -2068,7 +2069,13 @@ namespace {
             return resolvedTy;
         }
 
-        if (auto contextualType = CS.getContextualType(closure)) {
+        // Because we are only pulling out the result type from the contextual
+        // type, we avoid prematurely converting any inferrable types by setting
+        // forConstraint=false. Later on in inferClosureType we call
+        // replaceInferableTypesWithTypeVars before returning to ensure we don't
+        // introduce any placeholders into the constraint system.
+        if (auto contextualType =
+                CS.getContextualType(closure, /*forConstraint=*/false)) {
```

Comment:
I think you can take closure used to compute result type and turn it into a callback for `inferClosureType` instead.

Reply:
Or better only `getContextualType` could be a callback.

Reply:
I'm not sure how that helps us with the splitting of parameter and result types since we still won't use any type variables from the parameter contextual type in the system. E.g., in:

```
let f: (_) -> _ = { (x: Int) in 0.0 }
```

we will request the constraint contextual type in order to hook up the result type to the closure, but then `getContextualType` will also generate a type variable for the parameter type here, which _doesn't_ get used by `inferClosureType`. So the param type sticks around as an orphan type var and compilation fails with "unable to infer closure type in the current context".

Perhaps I am misunderstanding your callback suggestion?

Reply:
I'm proposing something like this:

```
FunctionType *inferClosureType(ClosureExpr *closure, 
                                                        llvm::function_ref<FunctionType *()> getContextualType) {
       ...

       Type resultTy = [&] {
           // checks for explicit type
          
           // No explicit result type, let's see if there is a contextual one
           if (auto contextualType = getContextualType()) {
             return contextualType->getResult();
           }
           ...
      }();
}
```

At the call site:

```
inferClosureType(closure, [&] {
   if (getContextualType(closure, /*forConstraint=*/false)->is<FunctionType>()) {
      return getContextualType(closure, /*forConstraint=*/true)->castTo<FunctionType>();
   }
   return nullptr;
});
```

Reply:
I think in your example opening would have to happen anyway because contextual type would be used in a contextual conversion constraint.

Reply:
Ugh, this is really-really bad but okay let's keep it the way it is.